prow/tide/contextregister.go

```diff
@@ -0,0 +1,145 @@
+	"sync"
+
+	"github.com/shurcooL/githubql"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/test-infra/prow/config"
+)
+
+type contextChecker interface {
+	// ignoreContext tells whether a context is optional.
+	ignoreContext(context Context) bool
+	// missingRequiredContexts tells if required contexts are missing from the list of contexts provided.
+	missingRequiredContexts([]Context) []Context
+}
+
+// newExpectedContext creates a Context with Expected state.
+// This should not only be used when contexts are missing.
+func newExpectedContext(c string) Context {
+	return Context{
+		Context:     githubql.String(c),
+		State:       githubql.StatusStateExpected,
+		Description: githubql.String(""),
+	}
+}
+
+// contextRegister implements contextChecker and allow registering of required and optional contexts.
+type contextRegister struct {
+	lock                sync.RWMutex
+	required, optional  sets.String
+	skipUnknownContexts bool
+}
+
+// newContextRegister instantiates a new contextRegister and register tide as optional by default
+// and uses Prow Config to find optional and required tests, as well as skipUnknownContexts.
+func newContextRegister(skipUnknownContexts bool) *contextRegister {
+	r := &contextRegister{
+		required:            sets.NewString(),
+		optional:            sets.NewString(),
+		skipUnknownContexts: skipUnknownContexts,
+	}
+	r.registerOptionalContexts(statusContext)
+	return r
+}
+
+// newContextRegisterFromPolicy instantiates a new contextRegister and register tide as optional by default
+// and uses Prow Config to find optional and required tests, as well as skipUnknownContexts.
+func newContextRegisterFromPolicy(policy *config.Policy) *contextRegister {
+	r := newContextRegister(false)
+	if policy != nil {
+		if policy.SkipUnknownContexts != nil {
+			r.skipUnknownContexts = *policy.SkipUnknownContexts
+		}
+		if policy.Protect != nil && *policy.Protect {
+			r.registerRequiredContexts(policy.Contexts...)
+		} else if r.skipUnknownContexts {
+			r.registerRequiredContexts(policy.Contexts...)
+		}
+
+	}
+	return r
+}
+
+// newContextRegisterFromPolicy instantiates a new contextRegister and register tide as optional by default
+// and uses Prow Config to find optional and required tests, as well as skipUnknownContexts.
+func newContextRegisterFromConfig(org, repo, branch string, c *config.Config) (*contextRegister, error) {
+	policy, err := c.GetBranchProtection(org, repo, branch)
+	if err != nil {
+		return nil, err
+	}
+	_, optional := config.BranchRequirements(org, repo, branch, c.Presubmits)
+	r := newContextRegisterFromPolicy(policy)
+	r.registerOptionalContexts(optional...)
+	return r, nil
+}
+
+// ignoreContext checks whether a context can be ignored.
+// Will return true if
+// - context is registered as optional
+// - required contexts are registered and the context provided is not required
+// Will return false otherwise. Every context is required.
+func (r *contextRegister) ignoreContext(c Context) bool {
```

Comment:
Shouldn't we also ignore the `tide` context regardless of whether or not it's required?

Reply:
we do, that s tide status. its registered by default. This implementation could be use in another context, so I did not want to hard code in there.